### PR TITLE
SQL transport purges expired messages

### PR DESCRIPTION
**Pull request by MarcinHoppe**
This is doco for Particular/NServiceBus.SqlServer#165.

### CODE REVIEW

**Comment by SimonCropp** [comment]
@MarcinHoppe is this a WIP until 165 has been deployed?

**Comment by MarcinHoppe** [comment]
@SimonCropp Yes. I marked this one as WIP. Good catch.

**Comment by SimonCropp** [comment]
@MarcinHoppe ok with that i mind does the doco need any "version this/version that" clarifications?

**Comment by MarcinHoppe** [comment]
@SimonCropp No, as long as we don't want to differentiate between minor versions. Do we?

V2 and V3 of the transport will work in the same way.

**Comment by SimonCropp** [comment]
@MarcinHoppe "as long as we don't want to differentiate between minor versions" yep that is fair enough. just wanted to check. thanks

File: nservicebus/sqlserver/design.md [hunk 1]
```diff
@@ -49,6 +49,10 @@ CREATE TABLE [schema].[queuename](
 CREATE CLUSTERED INDEX [Index_RowVersion] ON [schema].[queuename](
 	[RowVersion] ASC
 ) WITH (PAD_INDEX  = OFF, STATISTICS_NORECOMPUTE  = OFF, SORT_IN_TEMPDB = OFF, IGNORE_DUP_KEY = OFF, DROP_EXISTING = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS  = ON, ALLOW_PAGE_LOCKS  = ON) ON [PRIMARY]
+
+CREATE NONCLUSTERED INDEX [Index_Expires] ON [schema].[queuename](
```
Comment (tmasternak) [inline nservicebus/sqlserver/design.md]: I think we need to specify differences between version. Index should not exists for V1 for example

Comment (MarcinHoppe) [inline nservicebus/sqlserver/design.md]: Clarified (added Indexes) section.

**Comment by tmasternak** [comment]
Looks good to me. :+1:

**Comment by MarcinHoppe** [comment]
@Particular/docs-maintainers The SQL Server transport change described in this doco PR has been released. I made minor modifications but otherwise this is ready for review and merge.

File: nservicebus/sqlserver/design.md [hunk 2]
```diff
@@ -56,6 +69,13 @@ In version 2 the columns are directly mapped to the properties of `NServiceBus.T
 The tables are created by [installers](/nservicebus/operations/installers.md) when the application is started for the first time. It is required that the user account under which the installation of the host is performed has `CREATE TABLE` as well as `VIEW DEFINITION` permissions on the database in which the queues are to be created. The account under which the service runs does not have to have these permissions. Standard read/write/delete permissions (e.g. being member of `db_datawriter` and `db_datareader` roles) are enough.
 
 
+### Indexes
+
+Each queue table has a clustered index on the `[RowVersion]` column in order to speed up receiving messages from the queue table.
+
+Starting from version 2.2.2, each queue table also has an additional non-clustered index on the `[Expires]` column to speed up purging expired messages from the queue table. If the SQL Server transport discovers that a required index is missing, it will log an appropriate warning.
```
Comment (SimonCropp) [inline nservicebus/sqlserver/design.md]: if i am user running earlier than 2.2.2 what script do i run to patch this and add the new index?

Comment (SimonCropp) [inline nservicebus/sqlserver/design.md]: will the warning include the sql i need to run?

Comment (MarcinHoppe) [inline nservicebus/sqlserver/design.md]: No, the warning will just instruct the reader to consult the doco. The SQL to create the index is given in the snippet above and in the GG announcement. Do you think we need to spell it out explicitly?

When this feature / bug fix is ported properly to V3, I will add a section on how to create the index to the upgrade guide to aid users upgrading from SQL Server transport 2.2.1 (and earlier) directly to V3.

Comment (SimonCropp) [inline nservicebus/sqlserver/design.md]: the GG announcement is transient and does not reach everyone. so yes i think it needs to be called out.

Comment (MarcinHoppe) [inline nservicebus/sqlserver/design.md]: Fixed.

Comment (SimonCropp) [inline nservicebus/sqlserver/design.md]: perhaps 

> Starting from version 2.2.2, each queue table also has an additional non-clustered index on the `[Expires]` column. This index speeds up the purging expired messages from the queue table.

File: nservicebus/messaging/discard-old-messages.md [hunk 3]
```diff
@@ -52,4 +52,4 @@ The Azure transports only evaluate the TimeToBeReceived attribute for a message
 
 ### SQL transport
 
-The SQL transport evaluates the TimeToBeReceived attribute for a message when the message is received from the queue. Expired messages are not removed from the queue and their disk space will not be reclaimed until they reach the front of the queue and a consumer tries to read them. Using TimeToBeReceived as a disk saving measure on the SQL transports is not a great choice for queues with long-lived messages like audit and forward.
\ No newline at end of file
+The SQL transport runs a periodic task that removes expired messages from the queue. The task is first executed when the endpoint starts and is subsequently scheduled to execute 5 minutes after the previous run of the task has been completed. To promptly discard expired messages that arrive at the front of the queue between scheduled executions of the purging task, the transport also evaluates the TimeToBeReceived attribute for a message when the message is received from the queue.
```
Comment (SimonCropp) [inline nservicebus/messaging/discard-old-messages.md]: > To promptly discard expired messages that arrive at the front of the queue between scheduled executions of the purging task, the transport also evaluates the TimeToBeReceived attribute for a message when the message is received from the queue.

can this sentence be broken down ?

File: nservicebus/sqlserver/design.md (hunk 2 above)
Comment (MarcinHoppe) [inline nservicebus/sqlserver/design.md]: Fixed.

File: nservicebus/messaging/discard-old-messages.md (hunk 3 above)
Comment (MarcinHoppe) [inline nservicebus/messaging/discard-old-messages.md]: Fixed.

**Pull request merged**